src/contracts/protocol/libraries/logic/LiquidationLogic.sol

```diff
@@ -53,27 +62,126 @@ library LiquidationLogic {
+   * This parameter is inferred from MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD as the logic is dependent.
+   * Assuming a MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD of `n` a liquidation of `n+1` might result in `n/2` leftover which is assumed to be still economically liquidatable.
+   * This mechanic was introduced to ensure liquidators don't optimize gas by leaving some wei on the liquidation.
+   */
+  uint256 public constant MIN_LEFTOVER_BASE = MIN_BASE_MAX_CLOSE_FACTOR_THRESHOLD / 2;
+
+  /**
+   * @notice Reduces a portion or all of the deficit of a specified reserve by burning:
+   * - the equivalent aToken `amount` for assets with virtual accounting enabled
+   * - the equivalent `amount` of underlying for assets with virtual accounting disabled (e.g. GHO)
+   * The caller of this method MUST always be the Umbrella contract and the Umbrella contract is assumed to never have debt.
+   * @dev Emits the `DeficitCovered() event`.
+   * @dev If the coverage admin covers its entire balance, `ReserveUsedAsCollateralDisabled()` is emitted.
+   * @param reservesData The state of all the reserves
+   * @param userConfig The user configuration mapping that tracks the supplied/borrowed assets
+   * @param params The additional parameters needed to execute the eliminateDeficit function
+   */
+  function executeEliminateDeficit(
+    mapping(address => DataTypes.ReserveData) storage reservesData,
+    DataTypes.UserConfigurationMap storage userConfig,
+    DataTypes.ExecuteEliminateDeficitParams memory params
+  ) external {
+    require(params.amount != 0, Errors.INVALID_AMOUNT);
+
+    DataTypes.ReserveData storage reserve = reservesData[params.asset];
+    uint256 currentDeficit = reserve.deficit;
+
+    require(currentDeficit != 0, Errors.RESERVE_NOT_IN_DEFICIT);
+    require(!userConfig.isBorrowingAny(), Errors.USER_CANNOT_HAVE_DEBT);
+
+    DataTypes.ReserveCache memory reserveCache = reserve.cache();
+    reserve.updateState(reserveCache);
+    bool isActive = reserveCache.reserveConfiguration.getActive();
+    require(isActive, Errors.RESERVE_INACTIVE);
+
+    uint256 balanceWriteOff = params.amount;
+
+    if (params.amount > currentDeficit) {
+      balanceWriteOff = currentDeficit;
+    }
+
+    uint256 userBalance = reserveCache.reserveConfiguration.getIsVirtualAccActive()
+      ? IAToken(reserveCache.aTokenAddress).scaledBalanceOf(msg.sender).rayMul(
+        reserveCache.nextLiquidityIndex
+      )
+      : IERC20(params.asset).balanceOf(msg.sender);
+    require(balanceWriteOff <= userBalance, Errors.NOT_ENOUGH_AVAILABLE_USER_BALANCE);
+
+    if (reserveCache.reserveConfiguration.getIsVirtualAccActive()) {
+      // assets without virtual accounting can never be a collateral
+      bool isCollateral = userConfig.isUsingAsCollateral(reserve.id);
+      if (isCollateral && balanceWriteOff == userBalance) {
+        userConfig.setUsingAsCollateral(reserve.id, false);
```

Comment:
have we thought about imposing Umbrella not having the aTokens being used as collateral? Maybe we can even replace the `isBorrowingAny` with this check, and simplify this part.

Reply:
not sure I follow the rationale. My point was about imposing a specific state for Umbrella and do not see how being mutable relates.